File: src/apps/scala/slate-examples/src/main/scala/slate/examples/common/UserApi.scala
/**
<slate_header>
  author: Kishore Reddy
  url: https://github.com/kishorereddy/scala-slate
  copyright: 2015 Kishore Reddy
  license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  desc: a scala micro-framework
  usage: Please refer to license on github for more info.
</slate_header>
  */

package slate.examples.common

import slate.common.results.ResultSupportIn
import slate.common.{DateTime, Result}
import slate.core.apis._

@Api(area = "app", name = "users", desc = "api to access and manage users 3",
  roles= "@admin", auth = "app", protocol = "*")
class UserApi extends ApiBaseEntity[User] with ResultSupportIn
{
  var user = new User

  @ApiAction(name = "", desc = "activates a users account 3", roles= "@parent")
  @ApiArg("phone", "the phone number", true, "1-234-567-8901")
  def info(format:String = "json"): Result[Boolean] =
  {
    ok(Some(s"info"))
  }


  @ApiAction(name = "", desc = "activates a users account 3", roles= "@parent")
  @ApiArg("phone", "the phone number", true, "1-234-567-8901")
  def activate(phone:String, code:Int, isPremiumUser:Boolean, date:DateTime): Result[Boolean] =
  {
    ok(Some(s"activated $phone, $code, $isPremiumUser, $date"))
  }


  @ApiAction(name = "", desc = "invites a new user", roles= "@parent")
  //@ApiArg(name = "email"    , desc = "the email of invitee", required = true, eg = "johndoe@gmail.com")
  //@ApiArg(name = "phone"    , desc = "the phone of invitee", required = true, eg = "12345678901")
  //@ApiArg(name = "promoCode", desc = "promotion code"      , required = true, eg = "abc")
  def invite(email:String, phone:String, promoCode:String): Result[Boolean] =
  {
    ok(Some(s"sent invitation to $email, $phone, $promoCode"))
  }



  @ApiAction(name = "", desc = "invites a new user", roles= "@parent")
  //@ApiArg(name = "email"    , desc = "the email of invitee", required = true, eg = "johndoe@gmail.com")
  //@ApiArg(name = "phone"    , desc = "the phone of invitee", required = true, eg = "12345678901")
  //@ApiArg(name = "promoCode", desc = "promotion code"      , required = true, eg = "abc")
  def register(user:User): Result[Boolean] =
  {
    ok(Some(s"object user"))
  }



  //@ApiAction(desc= "creates a new user", roles= "@parent", args= "[" +
  //  "{ name = 'email' , desc = 'the email of invitee', required = true, eg = 'johndoe@gmail.com'}," +
  //  "{ name = 'first' , desc = 'the first name'      , required = true, eg = 'john'             }," +
  //  "{ name = 'last'  , desc = 'the last name'       , required = true, eg = 'doe'              }," +
  //  "{ name = 'isMale', desc = 'whether user is male', required = true, eg = 'true|false'       }," +
  //  "{ name = 'age'   , desc = 'age of the user'     , required = true, eg = '35'               })" +
  //  "]")
  def create(email:String, first:String, last:String, isMale:Boolean, age:Int): User =
  {
    user.email = email
    user.firstName = first
    user.lastName = last
    user.isMale = isMale
    user.age = age
    user
  }
}

File: src/lib/kotlin/slatekit-examples/src/main/kotlin/slatekit/examples/common/UserApi.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.examples.common

import slatekit.apis.Api
import slatekit.apis.Action
import slatekit.apis.Arg
import slatekit.apis.setup.AuthModes
import slatekit.apis.setup.Protocols
import slatekit.apis.setup.Verbs
import slatekit.common.*
import slatekit.common.auth.Roles
import slatekit.integration.common.ApiBaseEntity
import slatekit.integration.common.AppEntContext
import slatekit.results.Notice
import slatekit.results.Success
import slatekit.results.Try


@Api(area = "app", name = "users", desc = "api to access and manage users 3",
        auth = AuthModes.token, roles = Roles.all, verb = Verbs.auto, protocol = Protocols.all)
class UserApi(ctx: AppEntContext) : ApiBaseEntity<Long, User, UserService>(ctx, Long::class, User::class, UserService(ctx.ent, ctx.ent.getRepo(User::class))) {

  @Action(name = "", desc = "activates a users account 3", roles= "@parent")
  fun activate(phone:String, code:Int, isPremiumUser:Boolean, date: DateTime): Try<String> {
    return Success("activated $phone, $code, $isPremiumUser, $date")
  }

  var _user = User()


  /**
   * The api action is used to tell the slatekit system to make this method
   * available as an api action.
   * NOTE: The Arg annotation is OPTIONAL
   */
  @Action(name = "", desc = "activates a users account 3", roles= "@parent")
  @Arg("phone", "the phone number", true, "1-234-567-8901")
  fun info(format:String = "json"): Notice<Boolean> {
    return Success(true,"info")
  }


  @Action(name = "", desc = "invites a new user", roles= "@parent")
  @Arg(name = "email"    , desc = "the email of invitee", required = true, eg = "johndoe@gmail.com")
  @Arg(name = "phone"    , desc = "the phone of invitee", required = true, eg = "12345678901")
  @Arg(name = "promoCode", desc = "promotion code"      , required = true, eg = "abc")
  fun invite(email:String, phone:String, promoCode:String): Notice<Boolean> {
    return Success(true,"sent invitation to $email, $phone, $promoCode")
  }


  @Action(name = "", desc = "invites a new user", roles= "@parent")
  @Arg(name = "email"    , desc = "the email of invitee", required = true, eg = "johndoe@gmail.com")
  @Arg(name = "phone"    , desc = "the phone of invitee", required = true, eg = "12345678901")
  @Arg(name = "promoCode", desc = "promotion code"      , required = true, eg = "abc")
  fun register(user:User): Notice<Boolean> {
    return Success(true, "object user")
  }


  @Action(name = "", desc= "creates a new user", roles= "@parent" )
  @Arg(name = "email" , desc = "the email of invitee", required = true, eg = "johndoe@gmail.com" )
  @Arg(name = "first" , desc = "the first api"      , required = true, eg = "john"              )
  @Arg(name = "last"  , desc = "the last api"       , required = true, eg = "doe"               )
  @Arg(name = "isMale", desc = "whether user is male", required = true, eg = "true|false"        )
  @Arg(name = "age"   , desc = "age of the user"     , required = true, eg = "35"                )
  fun create(email:String, first:String, last:String, isMale:Boolean, age:Int): User {

    _user = _user.copy(email = email,
              firstName = first,
              lastName = last,
              isMale = isMale,
              age = age
    )
    return _user.copy()
  }
}
